### PR TITLE
Add some events to swarm/node/services

### PR DESCRIPTION
**Pull request by vdemeester**
Update cluster package (swarm mode), to generate some events 🐳. This is related to #23827. `vendor` is expected to build – will do an `engine-api` PR once in code-review :wink: .
- [ ] Possibily add more events ? more fields more some of those events ?
- [ ] Add integration tests
- [ ] Add documentation tests

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @vieux

**Comment by tonistiigi** [comment]
I don't think it makes much sense to add these events like that, except maybe init/join/leave if they are useful. If we add events for services and nodes, these should listen the storage changes directly from raft. With the current changes I don't see how these events could be useful as anyone subscribing to them would miss all the changes made by other managers, changes made by swarmkit internally or changes that are made directly in swarkit api.

**Comment by vdemeester** [comment]
> With the current changes I don't see how these events could be useful as anyone subscribing to them would miss all the changes made by other managers, changes made by swarmkit internally or changes that are made directly in swarkit api.

True :+1:. I'll update to only add the swarm init/join/leave in that PR and look at how to "listen the storage changes directly from raft".

File: daemon/events.go [hunk 1]
```diff
@@ -94,6 +94,17 @@ func (daemon *Daemon) LogDaemonEventWithAttributes(action string, attributes map
 	}
 }
 
+// LogEventWithAttributes generates an generic event with specific given attributes
```
Comment (allencloud) [inline daemon/events.go]: `an generic event` -> `a generic event`

File: vendor/src/github.com/docker/engine-api/types/events/events.go [hunk 2]
```diff
@@ -9,8 +9,14 @@ const (
 	VolumeEventType = "volume"
 	// NetworkEventType is the event type that networks generate
 	NetworkEventType = "network"
-	// DaemonEventType is the event type that daemon generate
+	// DaemonEventType is the event type that daemon generates
 	DaemonEventType = "daemon"
+	// SwarmEventType is the event type that swarm generates
+	SwarmEventType = "swarm"
```
Comment (allencloud) [inline vendor/src/github.com/docker/engine-api/types/events/events.go]: @vdemeester 
While this change is in the vendor of docker, And I have not found PRs in docker/engine-api repo.
May I have your thought on how to plan the procedure?

Comment (vdemeester) [inline vendor/src/github.com/docker/engine-api/types/events/events.go]: @allencloud I'm waiting for this PR to make it to `code-review` before doing a PR in engine-api (that's what I tend to do all the time :angel:)

**Comment by thaJeztah** [comment]
ping @vieux @tonistiigi PTAL

File: vendor/src/github.com/docker/engine-api/types/events/events.go [hunk 3]
```diff
@@ -9,8 +9,14 @@ const (
...
+	// NodeEventType is the event type that nodes generate
+	NodeEventType = "node"
+	// ServiceEventType is the event type that services generate
+	ServiceEventType = "service"
```
Comment (jhorwit2) [inline vendor/src/github.com/docker/engine-api/types/events/events.go]: Do you plan to publish all events related to a service? What i was hoping for was the ability to stream all events related to a service (create/update/containers kill/start/health change/etc). If there are 2 types (container & service) then the current way events are filtered won't be able to do show both.

**Comment by cleishm** [comment]
@tonistiigi - the issue #23827 was raised due to a desire to observe changes to confirmed swarm size (i.e. changes to the scale factor applied to a specific service). Does this constitute as "storage changes directly from raft"? If so, is there another issue that will address this use-case?

**Comment by tonistiigi** [comment]
@cleishm Yes, all swarm state is stored in raft including services, tasks, nodes etc. https://github.com/docker/swarmkit/issues/491 is a swarmkit issue about this.

**Comment by vdemeester** [comment]
Let's close this for now as most of the event should be provided by swarmkit/raft (and thus tracking of this feature is https://github.com/docker/swarmkit/issues/491 now) — and do not see too much use-cases for the `swarm init` event.

Please feel free to comment if you feel there is some use-cases 👼

**Pull request closed without merging**